cmd/azure.go

```diff
@@ -0,0 +1,333 @@
+			ContentType:     object.Properties.ContentType,
+			ContentEncoding: object.Properties.ContentEncoding,
+		})
+	}
+	result.Prefixes = resp.BlobPrefixes
+	return result, nil
+}
+
+// GetObject - Use Azure equivalent GetBlobRange.
+func (a AzureObjects) GetObject(bucket, object string, startOffset int64, length int64, writer io.Writer) (err error) {
+	byteRange := fmt.Sprintf("%d-%d", startOffset, startOffset+length-1)
+	if length == -1 {
+		byteRange = fmt.Sprintf("%d-", startOffset)
+	}
+	rc, err := a.client.GetBlobRange(bucket, object, byteRange, nil)
+	if err != nil {
+		return traceError(azureToObjectError(err, bucket, object))
+	}
+	io.Copy(writer, rc)
+	rc.Close()
+	return nil
+}
+
+// GetObjectInfo - Use Azure equivalent GetBlobProperties.
+func (a AzureObjects) GetObjectInfo(bucket, object string) (objInfo ObjectInfo, err error) {
+	prop, err := a.client.GetBlobProperties(bucket, object)
+	if err != nil {
+		return objInfo, traceError(azureToObjectError(err, bucket, object))
+	}
+	t, err := time.Parse(time.RFC1123, prop.LastModified)
+	if err != nil {
+		return objInfo, traceError(err)
+	}
+	objInfo.Bucket = bucket
+	objInfo.ContentEncoding = prop.ContentEncoding
+	objInfo.ContentType = prop.ContentType
+	objInfo.MD5Sum = prop.Etag
+	objInfo.ModTime = t
+	objInfo.Name = object
+	objInfo.Size = prop.ContentLength
+	return
+}
+
+// PutObject - Use Azure equivalent CreateBlockBlobFromReader.
+func (a AzureObjects) PutObject(bucket, object string, size int64, data io.Reader, metadata map[string]string, sha256sum string) (objInfo ObjectInfo, err error) {
+	err = a.client.CreateBlockBlobFromReader(bucket, object, uint64(size), data, nil)
+	if err != nil {
+		return objInfo, traceError(err)
+	}
+	return a.GetObjectInfo(bucket, object)
+}
+
+// CopyObject - Use Azure equivalent CopyBlob.
+func (a AzureObjects) CopyObject(srcBucket, srcObject, destBucket, destObject string, metadata map[string]string) (objInfo ObjectInfo, err error) {
+	err = a.client.CopyBlob(destBucket, destObject, a.client.GetBlobURL(srcBucket, srcObject))
+	if err != nil {
+		return objInfo, traceError(azureToObjectError(err, srcBucket, srcObject))
+	}
+	return a.GetObjectInfo(destBucket, destObject)
+}
+
+// DeleteObject - Use azure equivalent DeleteBlob.
+func (a AzureObjects) DeleteObject(bucket, object string) error {
+	err := a.client.DeleteBlob(bucket, object, nil)
+	if err != nil {
+		return traceError(azureToObjectError(err, bucket, object))
+	}
+
+	return nil
+}
+
+// ListMultipartUploads - Incomplete implementation, for now just return the prefix if it is an incomplete upload.
+func (a AzureObjects) ListMultipartUploads(bucket, prefix, keyMarker, uploadIDMarker, delimiter string, maxUploads int) (result ListMultipartsInfo, err error) {
+	result.MaxUploads = maxUploads
+	result.Prefix = prefix
+	result.Delimiter = delimiter
+	resp, err := a.ListObjectParts(bucket, prefix, prefix, 1, 1000)
+	if err != nil {
+		return result, nil
```

Comment:
In case ListObjectParts returns error, it would mean that no such incomplete upload exists on azure storage - in which case we return nil. This plays a role for mc and our client libs. i.e client does ListMultipartUploads to figure the previous uploadID to continue uploading from where it left off. If we return error the upload never starts and always returns error.